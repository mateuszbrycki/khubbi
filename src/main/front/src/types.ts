import {DateTimeFormatter, LocalDateTime, ZonedDateTime, ZoneId} from "js-joda";

export class EventDate {
    value: ZonedDateTime

    public constructor(value: ZonedDateTime) {
        this.value = value
    }

    public toString(): string {
        return this.value.toString();
    }

    public format(format: string): string {
        return this.value.format(DateTimeFormatter.ofPattern(format))
    }

    public toISODateTime(): string {
        return this.value.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
    }

    public static now(): EventDate {
        return new EventDate(ZonedDateTime.now());
    }

    public static ofDateAndTime(date: string): EventDate {
        return new EventDate(LocalDateTime.parse(date)
            .atZone(ZoneId.UTC));
    }

}

export interface Event {
    readonly note: string;
    readonly date: EventDate;
    readonly id: string;
}

export interface User {
    readonly id: string;
    readonly email: string;
    readonly createdAt: string;
}

export class EventForms {
    public static readonly NOTE = new EventForms('Note');
    public static readonly PICTURE = new EventForms('Picture');
    public static readonly EVENT = new EventForms('Event');
    public static readonly EXPENSE = new EventForms('Expense');
    public static readonly REMINDER = new EventForms('Reminder');

    private constructor(public readonly name: string) {
    }
}