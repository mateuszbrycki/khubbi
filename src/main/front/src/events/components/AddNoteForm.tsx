import React, {useState} from "react";
import {Col, Row} from "react-bootstrap";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import {EventDate} from "../../types";

export const DATE_FORMAT = "yyyy-MM-dd";
export const TIME_FORMAT = "HH:mm";

export interface AddNoteFormProps {
}

export interface AddNoteFormActionProps {
    readonly addNote: (note: string, date: EventDate) => void
    readonly currentDateProvider: () => EventDate
}

const AddNoteForm: React.FC<AddNoteFormProps & AddNoteFormActionProps> = (props) => {

    const {addNote, currentDateProvider} = props

    const [note, setNote] = useState<string>("")
    const [date, setDate] = useState<string>(currentDateProvider().format(DATE_FORMAT))
    const [time, setTime] = useState<string>(currentDateProvider().format(TIME_FORMAT))

    const resetFormValues = () => {
        setNote("");
        setDate(currentDateProvider().format(DATE_FORMAT));
        setTime(currentDateProvider().format(TIME_FORMAT));
    }

    return <Form onSubmit={event => {
        addNote(note, EventDate.ofDateAndTime(`${date}T${time}`));
        event.preventDefault();
        resetFormValues();
    }}>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="note">
                <Form.Control as="textarea"
                              data-testid="note-textarea"
                              rows={5}
                              value={note}
                              onChange={(event) => setNote(event.target.value)}
                />
            </Form.Group>
        </Row>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="noteAddDate">
                <Form.Control type="date"
                              data-testid="note-add-date"
                              value={date}
                              onChange={(event) => setDate(event.target.value)}
                />

            </Form.Group>
            <Form.Group as={Col} controlId="noteAddTime">
                <Form.Control type="time"
                              data-testid="note-add-time"
                              value={time}
                              onChange={(event) => setTime(event.target.value)}
                />
            </Form.Group>
            <Form.Group as={Col} controlId="submitForm">
                <Button type="submit" data-testid="add-note-button">Add note</Button>
            </Form.Group>
        </Row>
    </Form>

}

export default AddNoteForm
