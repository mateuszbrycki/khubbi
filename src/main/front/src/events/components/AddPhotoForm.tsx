import React, {useState} from "react";
import {EventDate} from "../../types";
import {Col, Row} from "react-bootstrap";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import {DATE_FORMAT, TIME_FORMAT} from "./AddNoteForm";
import {DateTimeFormatter, Instant, LocalDateTime, ZonedDateTime, ZoneId, ZoneOffset} from "js-joda";

export interface AddPhotoFormProps {
}

export interface AddPhotoFormActionProps {
    readonly addPhoto: (description: string, photo: File, date: EventDate) => void
    readonly currentDateProvider: () => EventDate
}


const AddPhotoForm: React.FC<AddPhotoFormProps & AddPhotoFormActionProps> = (props) => {

    const {addPhoto, currentDateProvider} = props

    const [description, setDescription] = useState<string>("")
    const [file, setFile] = useState<File | null>(null)
    const [date, setDate] = useState<string>(currentDateProvider().format(DATE_FORMAT))
    const [time, setTime] = useState<string>(currentDateProvider().format(TIME_FORMAT))
    const [overridePhotoDate, setOverridePhotoDate] = useState<boolean>(false)

    const resetFormValues = () => {
        setDescription("");
        setDate(currentDateProvider().format(DATE_FORMAT));
        setTime(currentDateProvider().format(TIME_FORMAT));
        setFile(null);
        setOverridePhotoDate(false);
    }

    return <Form onSubmit={event => {
        if (file) {
            addPhoto(description, file, EventDate.ofDateAndTime(`${date}T${time}`));
            event.preventDefault();
            resetFormValues();
        }
    }}>
        <Row className="mb-3">
            <Form.Group controlId="formFile" className="mb-3">
                <Form.Control type="file"
                              data-testid="photo-file"
                              onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
                                  if (event.target.files) {
                                      const file: File = event.target.files[0]
                                      const fileModifiedDate: ZonedDateTime =
                                          ZonedDateTime.ofLocal(
                                              LocalDateTime.ofInstant(
                                                  Instant.ofEpochMilli(file.lastModified)), ZoneId.SYSTEM, null)

                                      setFile(file);
                                      setDate(fileModifiedDate.format(DateTimeFormatter.ofPattern(DATE_FORMAT)))
                                      setTime(fileModifiedDate.format(DateTimeFormatter.ofPattern(TIME_FORMAT)))
                                  }
                              }}
                              accept=".png,.jpg,.jpeg,.webp"
                />
            </Form.Group>
        </Row>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="description">
                <Form.Control as="textarea"
                              data-testid="photo-description"
                              rows={5}
                              value={description}
                              placeholder="Photo description"
                              onChange={(event) => setDescription(event.target.value)}
                />
            </Form.Group>
        </Row>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="overridePhotoDate">
                <Form.Check
                    checked={overridePhotoDate}
                    data-testid="photo-overridePhotoDate"
                    type="checkbox"
                    id="overridePhotoDate-checkbox"
                    label="Override photo date"
                    onChange={(event) => setOverridePhotoDate(event.target.checked)}
                />
            </Form.Group>
        </Row>
        <Row className="mb-3">
            <Form.Group as={Col} controlId="photoAddDate">
                <Form.Control type="date"
                              disabled={!overridePhotoDate}
                              data-testid="photo-add-date"
                              value={date}
                              onChange={(event) => setDate(event.target.value)}
                />

            </Form.Group>
            <Form.Group as={Col} controlId="photoAddTime">
                <Form.Control type="time"
                              disabled={!overridePhotoDate}
                              data-testid="photo-add-time"
                              value={time}
                              onChange={(event) => setTime(event.target.value)}
                />
            </Form.Group>

            <Form.Group as={Col} controlId="submitForm">
                <Button type="submit" data-testid="add-photo-button">Add photo</Button>
            </Form.Group>
        </Row>
    </Form>

}

export default AddPhotoForm