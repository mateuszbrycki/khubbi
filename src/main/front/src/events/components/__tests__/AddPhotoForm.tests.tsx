import ShallowRenderer from "react-test-renderer/shallow";
import AddPhotoForm from "../AddPhotoForm";
import {EventDate} from "../../../types";
import {act, fireEvent, render} from "@testing-library/react";
import {DATE_FORMAT, TIME_FORMAT} from "../AddNoteForm";

const testCurrentDateProvider = () => EventDate.ofDateAndTime("2024-07-12T20:00")
const addPhotoMock = jest.fn(() => Promise.resolve());
const file = new File(["(⌐□_□)"], "image.png", {type: "image/png", lastModified: 1722023600022});

test('should render form', () => {
    const renderer = ShallowRenderer.createRenderer()
    renderer.render(<AddPhotoForm addPhoto={addPhotoMock} currentDateProvider={testCurrentDateProvider}/>)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
})

test('should call add photo', async () => {
    const {getByTestId} = render(<AddPhotoForm addPhoto={addPhotoMock} currentDateProvider={testCurrentDateProvider}/>)

    await act(async () => {
        fireEvent.change(getByTestId("photo-file"), {target: {files: [file]}})
        fireEvent.change(getByTestId("photo-description"), {target: {value: "test-description"}})

        getByTestId("add-photo-button").click();
    })

    expect(addPhotoMock).toHaveBeenCalledWith("test-description", file, EventDate.ofDateAndTime("2024-07-26T21:53"));

})

test('should call add photo with override date', async () => {
    const {getByTestId} = render(<AddPhotoForm addPhoto={addPhotoMock} currentDateProvider={testCurrentDateProvider}/>)

    await act(async () => {
        fireEvent.change(getByTestId("photo-file"), {target: {files: [file]}})
        fireEvent.change(getByTestId("photo-description"), {target: {value: "test-description"}})
        fireEvent.change(getByTestId("photo-overridePhotoDate"), {target: {checked: true}})
        fireEvent.change(getByTestId("photo-add-date"), {target: {value: "2023-03-12"}})
        fireEvent.change(getByTestId("photo-add-time"), {target: {value: "18:47"}})

        getByTestId("add-photo-button").click();
    })

    expect(addPhotoMock).toHaveBeenCalledWith("test-description", file, EventDate.ofDateAndTime("2023-03-12T18:47"));

})

test('should reset form after submit', async () => {
    const {getByTestId} = render(<AddPhotoForm addPhoto={addPhotoMock} currentDateProvider={testCurrentDateProvider}/>)

    await act(async () => {
        fireEvent.change(getByTestId("photo-file"), {target: {files: [file]}})
        fireEvent.change(getByTestId("photo-description"), {target: {value: "test-description"}})
        fireEvent.change(getByTestId("photo-overridePhotoDate"), {target: {checked: true}})
        fireEvent.change(getByTestId("photo-add-date"), {target: {value: "2023-03-12"}})
        fireEvent.change(getByTestId("photo-add-time"), {target: {value: "18:47"}})

        getByTestId("add-photo-button").click();
    })

    expect(addPhotoMock).toHaveBeenCalledWith("test-description", file, EventDate.ofDateAndTime("2023-03-12T18:47"));

    expect(getByTestId("photo-add-date").getAttribute("value")).toEqual(testCurrentDateProvider().format(DATE_FORMAT));
    expect(getByTestId("photo-add-time").getAttribute("value")).toEqual(testCurrentDateProvider().format(TIME_FORMAT));
})