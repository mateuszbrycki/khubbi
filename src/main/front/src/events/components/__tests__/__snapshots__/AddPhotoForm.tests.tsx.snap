// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form 1`] = `
<Form
  onSubmit={[Function]}
>
  <Row
    className="mb-3"
  >
    <FormGroup
      className="mb-3"
      controlId="formFile"
    >
      <FormControl
        accept=".png,.jpg,.jpeg,.webp"
        data-testid="photo-file"
        onChange={[Function]}
        type="file"
      />
    </FormGroup>
  </Row>
  <Row
    className="mb-3"
  >
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="description"
    >
      <FormControl
        as="textarea"
        data-testid="photo-description"
        onChange={[Function]}
        placeholder="Photo description"
        rows={5}
        value=""
      />
    </FormGroup>
  </Row>
  <Row
    className="mb-3"
  >
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="overridePhotoDate"
    >
      <FormCheck
        checked={false}
        data-testid="photo-overridePhotoDate"
        id="overridePhotoDate-checkbox"
        label="Override photo date"
        onChange={[Function]}
        type="checkbox"
      />
    </FormGroup>
  </Row>
  <Row
    className="mb-3"
  >
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="photoAddDate"
    >
      <FormControl
        data-testid="photo-add-date"
        disabled={true}
        onChange={[Function]}
        type="date"
        value="2024-07-12"
      />
    </FormGroup>
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="photoAddTime"
    >
      <FormControl
        data-testid="photo-add-time"
        disabled={true}
        onChange={[Function]}
        type="time"
        value="20:00"
      />
    </FormGroup>
    <FormGroup
      as={
        {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      controlId="submitForm"
    >
      <Button
        data-testid="add-photo-button"
        type="submit"
      >
        Add photo
      </Button>
    </FormGroup>
  </Row>
</Form>
`;
